libnd4j/include/array/impl/ShapeDescriptor.cpp

```diff
@@ -724,16 +719,89 @@ ShapeDescriptor  * ShapeDescriptor::paddedBufferDescriptor(const DataType type,
 
 namespace std {
 size_t hash<sd::ShapeDescriptor>::operator()(sd::ShapeDescriptor k) const {
-  auto res = std::hash<char>()(k.order());
-  res ^= std::hash<int>()((int)k.dataType()) + 0x9e3779b9 + (res << 6) + (res >> 2);
-  sd::LongType * shape_strides = k.shape_strides();
-  auto ptr = shape_strides;
-  //dont include strides if its' ews==1
-  int stop =  k.rank() * 2;
-  for (int j = 0; j < stop; j++) {
-    res ^= std::hash<sd::LongType>()(ptr[j]) + 0x9e3779b9 + (res << 6) + (res >> 2);
+  // Check cache first
```

Comment:
You've got the hash implementation in a couple of places, if it can be reused, try to reuse it.